src/pyhf/constraints.py

```diff
@@ -38,85 +44,105 @@ def _precompute(self):
+
+            _normal_sigmas = default_backend.concatenate(normal_constraint_sigmas)
+            sigmas = default_backend.reshape(_normal_sigmas, (1, -1))  # (1, normals)
+            self._batched_sigmas = default_backend.tile(
+                sigmas, (self.batch_size or 1, 1)
             )
-            normal_data = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_data,
-                    )
+
+            # self.parameter_helper.index_selection  is [ (batch, parslice) ] ]
+            access_field = None
+            for x in self.parameter_helper.index_selection:
+                access_field = (
+                    x
+                    if access_field is None
+                    else default_backend.concatenate([access_field, x], axis=1)
                 )
-            )
+            # access field is (nbatch, normals)
+            self._access_field = access_field
 
-            self.normal_data = tensorlib.astensor(
-                default_backend.tolist(normal_data), dtype='int'
-            )
-            self.normal_sigmas = tensorlib.astensor(
-                default_backend.tolist(normal_sigmas)
-            )
-            self.normal_mean_idc = tensorlib.astensor(
-                default_backend.tolist(normal_mean_idc), dtype='int'
-            )
         else:
-            self.normal_data, self.normal_sigmas, self.normal_mean_idc = (
-                None,
-                None,
-                None,
-            )
+            self._normal_data = None
+            self._batched_sigmas = None
+            self._access_field = None
+
+        self._precompute()
+        events.subscribe('tensorlib_changed')(self._precompute)
+
+    def _precompute(self):
+        if not self.parameter_helper.index_selection:
+            return
+        tensorlib, _ = get_backend()
+        self.batched_sigmas = tensorlib.astensor(self._batched_sigmas)
+        self.normal_data = tensorlib.astensor(self._normal_data, dtype='int')
+        self.access_field = tensorlib.astensor(self._access_field, dtype='int')
 
     def logpdf(self, auxdata, pars):
-        if self.normal_data is None:
-            return 0
         tensorlib, _ = get_backend()
+        if not self.parameter_helper.index_selection:
+            return (
+                tensorlib.zeros(self.batch_size)
+                if self.batch_size is not None
+                else tensorlib.astensor(0.0)[0]
+            )
+
+        pars = tensorlib.astensor(pars)
+        if self.batch_size == 1 or self.batch_size is None:
+            batched_pars = tensorlib.reshape(
+                pars, (self.batch_size or 1,) + tensorlib.shape(pars)
+            )
+        else:
+            batched_pars = pars
+
+        flat_pars = tensorlib.reshape(batched_pars, (-1,))
+        normal_means = tensorlib.gather(flat_pars, self.access_field)
```

Comment:
why `means`? This variable just seems to be `normal_pars`.

Reply:
yeah, the means of the normal constraint are set by the parameter directly but I think making this explicit is useful here

Reply:
"means" here refers to the mean value of the gaussian constraint term

Reply:
Right, it happens that in this case, the `mu` of the gaussian is the parameter value itself... I would prefer to change to `normal_pars` with a comment above of `# NB: this is also the mu of the gaussian constraint term`

Reply:
hm, I don't think that's much more cleaner. To me "normal_means" refers to the "means" or the normal distribution in this constraint computation. I'd lean towards keeping as is